src/toc.ts

```diff
@@ -2,9 +2,19 @@
 
 import * as path from 'path';
 import * as stringSimilarity from 'string-similarity';
+import * as vscode from 'vscode';
```

Comment:
This looks unnecessary. With the following line, we can save a lot of space for `vscode.`

Reply:
Recommended practice:

[**Use the namespace import pattern if you’re importing a large number of things.**](https://www.typescriptlang.org/docs/handbook/modules.html#use-the-namespace-import-pattern-if-youre-importing-a-large-number-of-things)

---

From my experience, when probably or already importing more than **five** items, the namespace import is convenient.

Of course, when reading commit history, I noticed you're hesitant about it.

Reply:
> Of course, when reading commit history, I noticed you're hesitant about it.

I gradually came to the current preference (saving space as the main criterion, it also improves clarity). You can easily get a very long function header if we keep all the `vscode.` (especially for some VS Code APIs). So in contrast, I prefer to use `vscode.`  if there are not many `vscode.` accesses in that file. (I remember there was a commit where I removed many `vscode.`.)

> Recommended practice:
>
> Use the namespace import pattern if you’re importing a large number of things.

It looks more like a thing that should be done on the VS Code API side. IMO, it is better to have ONE long line in the beginning than having to repeat `vscode.` many times everywhere.

Reply:
> I remember there was a commit where I removed many `vscode`.

You actually added and removed it many many times.

---

> It looks more like a thing that should be done on the VS Code API side.

?

This advice is about `import`. It is for **consumers** (people like us), not module creators.

---

> IMO, it is better to have ONE long line in the beginning

Preference again ... 😅

* I don't consider line length as a problem. In Android development, we write much longer.
* In .NET, I only do `using Namespace` for `System` plus a few well-known namespaces in the framework. For other cases, either `using alias = Namespace`, or fully qualified name. That makes me feel safe.
* From my experience, code completion is more handy with fully qualified name. TAB TAB TAB, done. Very fast, only very few candidates.

Reply:
> You actually added and removed it many many times.

Finally, I settled down with the current preference 😏.

> This advice is about `import`. It is for consumers (people like us), not module creators.

Alright, I misread that. It is probably good for engineering, but I still keep my personal taste (see below).

> Preference again ... 😅

Of course it is preference. And your text explained a lot about the preference diff. Java and C# are known to be "heavy"/"tedious" (esp. Java 😂). And my main working language is Python... Although I do not like many aspects of it (e.g. indentation), I have to admit its conciseness is a relief.

In the case, I'm okay with both (I have no intention to 100% align with the recommended practices), as long as we don't add `vscode.` back to some lines like this

https://github.com/yzhang-gh/vscode-markdown/blob/347059896d4b779c8d9459970f5ca07db1dfb043/src/util.ts#L23

(Three `vscode.` in a line, I cannot take it.)

Reply:
OK. Rolled back to named import in `toc.ts`.

But I'll still do it my way in **new** modules.